tests/includes/juju.sh

```diff
@@ -203,6 +203,11 @@ introspect_controller() {
 
     name=${1}
 
-    juju ssh -m controller 0 bash -lic "juju_engine_report" > "${TEST_DIR}/${name}-juju_engine_report.txt"
-    juju ssh -m controller 0 bash -lic "juju_goroutines" > "${TEST_DIR}/${name}-juju_goroutines.txt"
+    idents=$(juju machines -m ${name}:controller --format=json | jq ".machines | keys | .[]")
+    if [ -z "${idents}" ]; then
+        return
+    fi
+
+    echo "${idents}" | xargs -I % juju ssh -m "${name}:controller" % bash -lc "juju_engine_report" > "${TEST_DIR}/${name}-juju_engine_reports.txt"
```

Comment:
Shouldn't this work in append mode (i.e use `>>`) since you are potentially iterating multiple controllers?
Also, if TEST_DIR is re-usable you may first want to run something like this to reset them before the xargs calls:

```
echo > ${TEST_DIR}/${name}-juju_engine_reports.txt
echo > ${TEST_DIR}/${name}-juju_goroutines.txt
```

Reply:
Additionally, if you are concatenating them maybe you should also consider echoing a header line to indicate the machine you got the reports from.

Reply:
I did think that at first, but actually because we xargs, they considered one command, so get dumped as one.

`${TEST_DIR}` is just a tmp directory where artifacts from the test directory live or bundles/charms are when you pre-process them (see cmr-bundles test suite).